### ISSUE TITLE
Not All Compensation Is Made Equal

### ISSUE THREAD

**Issue by pavlovcik** [ubiquity/bounty-bot#221]
### Context
Some contributors collect a base pay/salary in the organization and have a hybrid compensation model being able to collect bounties. Assuming we are able to create three permission levels for users in the DevPool: 1. Treasurer role (currently just me at Ubiquity) 2. Repository/code maintainers (admins) and 3. Public members (bounty hunters)

The Treasurer roles should be able to configure the multiplier on the payouts per user.

Examples: 
1. If a contributor receives zero base pay/salary, then they should be eligible for the full amount for the bounties. 
2. If a contributor receives a mid base pay/salary, then the payouts should be cut in half.
3. If a contributor receives a high/max base pay/salary, then they should not be eligible for bounties.

When the payout link comment is posted, the bot should transparently describe the modifications (if any) in base pay.

Full disclosure, I pay myself a higher base pay/salary than the others in Ubiquity, but I don't collect any bounties even if I complete them. I think that setup makes more sense. On the other end of the spectrum, we have some public bounty hunters that I feel like could benefit from a pay boost for completing bounties (so we can go ahead and increase the baseRate for all the bounty price calculations!) 

My vision is one day to work out the compensation for issue creators, PR reviewers etc to also receive compensation, and gradually move everybody (including myself) off of base pay/salary.

### Spec
1. Build support for Treasurer role by leveraging the built in GitHub user role: `billing_manager`
2. Create a command ```/setPaymentMultiplier` @username float```
  - e.g. 
    - /setBountyMultiplier @pavlovcik 0
    - /setBountyMultiplier @0xcodercrane 0.5
3. Make the command only callable by `billing_manager` and `admin` roles.

**Comment by devpanther** [comment]
I'd need more details on this, since its related to the issue i'm working on @pavlovcik

**Comment by devpanther** [comment]
The bounty multiplier alters the price added on the tags or its just stored for the final payout?

**Comment by pavlovcik** [comment]
Just final payout. The tags should represent the full amount, which is what public bounty hunters would be eligible for. 

The people on salary would be aware of this rule and know their own multipliers so it shouldn't be an issue.

**Comment by devpanther** [comment]
And the `Treasurer` role, probably gonna be a list of users in a table right.. so only you or a group of persons can call

And the `/setBountyMultiplier`, is it called on an issue's comment or somewhere private since it takes the contributors username anyway

**Comment by pavlovcik** [comment]
The commands can be called on any issue the bot has access to. It will be public and only need to be set once somewhere.

**Comment by devpanther** [comment]
Makes sense to automatically track the treasurer doesn't top up the permit using this repo, do they?.. 

And how do we link the hex address back to the GitHub user in charge

> For the code maintainer level, we can ask the GitHub user API to see if the user is added on the organization as a code maintainer.

Yea, we can query the Github API for the user id and set their multiplier

**Comment by pavlovcik** [comment]
### User Permissions
I just learned that we can officially set five types of roles to each user in our DevPool using built in GitHub features. If somebody is added to an organization, the four roles they can have are: 
1. `admin`
2. `member`
3. `billing_manager`
4. `hiring_manager`

---

5. `public` If they are not added to the organization, then they are considered as a public bounty hunter with minimum permissions.

We can query this information from the GitHub API.

```
You can use the GitHub API to determine if a user is either a "member" or "owner" of an organization. You can do this by using the "List your organization memberships" endpoint, which returns a list of all the organizations that the authenticated user is a member of, along with their role in each organization.

Here's an example API request you could make:

GET /user/orgs
This endpoint will return a JSON array of organization objects, which includes a "role" property that indicates whether the user is an "owner" or "member" of each organization. 
```

We can use `billing_manager` role as the treasurer!

### Appendix

To clarify, the vision is as follows, note the separate user privilege/permission levels: 
- `public`: Bounty hunters can propose new bounties by creating a new issue. 
- `public`: The community (in particular the proposer) can provide the time estimate for the bounty (i.e. set time label)
- `member`: The repo/product/project manager sets the priority

If the bounty is created, and the time has been added; the way that it is "approved" or authorized for funding is when the project manager also assigns a priority. Otherwise, bounty hunters can propose bogus bounties and close them and cash out, draining the funds.

**Comment by devpanther** [comment]
Totally makes sense with this approach,

- So we've covered allowing only usernames present in any of these roles to be tagged in the command.

- `billing_admin` is treasurer and admin can call treasurers command too

- Price is reduced on final payout, labels remains unchanged

**Comment by pavlovcik** [comment]
> So we've covered allowing only usernames present in any of these roles to be tagged in the command.

I don't understand this point can you rephrase?

**Comment by devpanther** [comment]
> > So we've covered allowing only usernames present in any of these roles to be tagged in the command.
> 
> I don't understand this point can you rephrase?

I've updated the comment

**Comment by devpanther** [comment]
> We can query this information from the GitHub API.

It doesn't show up when the contributor isn't a member of the org..

**Comment by devpanther** [comment]
For example @0xcodercrane checked out because he's a member of this org but contributors doesn't

Perhaps this command is going to affect more of members than contributors

**Comment by pavlovcik** [comment]
> It doesn't show up when the contributor isn't a member of the org..

This is expected

> `public` If they are not added to the organization, then they are considered as a public bounty hunter with minimum permissions.

The other contributors are on the `public` permissions level.

**Comment by devpanther** [comment]
@pavlovcik do you think its okay to create a `multiplier` table or add it to the `address` table

**Comment by pavlovcik** [comment]
@0xcodercrane is going to know how to answer your question better. 

I'm still leaning towards using comments as our database, with the benefit being that we don't need to maintain a database but the con being that memory doesn't persist across repositories. Seems like the most interesting architecture but we eventually will need to persist analytics and performance data somewhere for internal use across all partner repos. 

@0xcodercrane what do you think?

**Comment by devpanther** [comment]
> I'm still leaning towards using comments as our database

Persisting this data among diff repo and issues will be an issue, even if it works.. It's going to take significant time finding where it was set

I think it worked really well for followup unassign watcher because it's per issues and it just loops through the comments. 

I've added the multiplier to the supabase table and I think data stored there now shouldn't be too much.. We only have addresses and the username.

**Comment by devpanther** [comment]
Maybe what'd we do for a refactor will be creating a separate issue or repo that's for setting multipliers and data only

If that would work, but I'd prefer a database

**Comment by 0xcodercrane** [comment]
I recommend the database because it doesn't require any effort from us for now. I don't think its a good idea to add it to `wallet` table.

**Comment by devpanther** [comment]
> Do you have any updates @devpanther? If you would like to release the bounty back to the DevPool, please comment `/unassign`

Ready to merge @pavlovcik

**Comment by pavlovcik** [comment]
@0xcodercrane I think you should merge it after you test that it's all working!

**Comment by devpanther** [comment]
> > Do you have any updates @devpanther? If you would like to release the bounty back to the DevPool, please comment `/unassign`
> 
> Ready to merge @pavlovcik @0xcodercrane

bump @pavlovcik @0xcodercrane, its been 2 weeks

**Comment by pavlovcik** [comment]
I think it looks good to merge. 0xcodercrane is the lead on this codebase though.
